File: java-springboot-js/src/main/java/com/payneteasy/hostedfields/Control.java
package com.payneteasy.hostedfields;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;
import java.util.Map;

/**
 * The checksum the gateway signs its callbacks with, kept apart from the routes so it can be
 * tested without a server.
 *
 * <p>https://doc.payneteasy.com/integration/API_commands/merchant_callback_parameters.html
 */
final class Control {

    private Control() {}

    /** {@code sha1(status + orderid + merchant_order + merchant_control)}, lowercase hex. */
    static String checksum(String status, String orderId, String merchantOrder, String merchantControl) {
        try {
            MessageDigest sha1 = MessageDigest.getInstance("SHA-1");
            byte[] digest =
                    sha1.digest((status + orderId + merchantOrder + merchantControl).getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(digest);
        } catch (NoSuchAlgorithmException e) {
            // SHA-1 is one of the digests every JVM is required to have
            throw new IllegalStateException(e);
        }
    }

    /**
     * Whether a callback really came from the gateway. A missing parameter reads as "", so an
     * empty callback still has a well-defined checksum rather than a special case.
     *
     * <p>{@code MessageDigest.isEqual} rather than {@code equals}: it does not stop at the first
     * differing byte, and it answers false for a length mismatch instead of throwing, so a forged
     * short control is a 403 and not a 500.
     */
    static boolean valid(Map<String, String> form, String merchantControl) {
        String expected = checksum(
                form.getOrDefault("status", ""),
                form.getOrDefault("orderid", ""),
                form.getOrDefault("merchant_order", ""),
                merchantControl);
        return MessageDigest.isEqual(
                expected.getBytes(StandardCharsets.UTF_8),
                form.getOrDefault("control", "").getBytes(StandardCharsets.UTF_8));
    }
}

File: kotlin-ktor-js/src/main/kotlin/com/payneteasy/hostedfields/Control.kt
package com.payneteasy.hostedfields

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.HexFormat

/**
 * The checksum the gateway signs its 3DS return with.
 *
 * It is checked twice over the same function: once on the POST the gateway makes to
 * `/result/callback`, and again on the query `GET /result` is reached with. The browser carries
 * the four values in between but cannot forge them, because it does not know MERCHANT_CONTROL.
 */
object Control {
    /** The four values the gateway signs, in the order it concatenates them. */
    val SIGNED_FIELDS = listOf("status", "orderid", "merchant_order", "control")

    /** `sha1(status + orderid + merchant_order + merchant_control)`, lowercase hex. */
    fun checksum(
        status: String,
        orderId: String,
        merchantOrder: String,
        merchantControl: String,
    ): String {
        val digest =
            MessageDigest
                .getInstance("SHA-1")
                .digest((status + orderId + merchantOrder + merchantControl).toByteArray(StandardCharsets.UTF_8))
        return HexFormat.of().formatHex(digest)
    }

    /**
     * Whether a callback or a result query carries the checksum it should.
     *
     * A missing field counts as empty, so an empty callback still has a well-defined checksum —
     * which is what stops a bare `/result?orderid=…` with no control from passing. The comparison
     * is constant-time and never throws on a control of the wrong length: that has to be a 403,
     * not an exception on the way to a 500.
     */
    fun valid(
        form: Map<String, String>,
        merchantControl: String,
    ): Boolean {
        val expected =
            checksum(
                form["status"].orEmpty(),
                form["orderid"].orEmpty(),
                form["merchant_order"].orEmpty(),
                merchantControl,
            )
        return MessageDigest.isEqual(
            expected.toByteArray(StandardCharsets.UTF_8),
            form["control"].orEmpty().toByteArray(StandardCharsets.UTF_8),
        )
    }
}
